In [1]:
#importing libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt # we only need pyplot
from sklearn.model_selection import train_test_split

In [2]:
#We load the dataset and look at the overall statistics such as mean, count, and median.

music = pd.read_csv("music_genre.csv")
music


instance_id           artist_name                track_name  \
0          32894.0              Röyksopp      Röyksopp's Night Out   
1          46652.0  Thievery Corporation          The Shining Path   
2          30097.0        Dillon Francis                 Hurricane   
3          62177.0              Dubloadz                     Nitro   
4          24907.0           What So Not          Divide & Conquer   
...            ...                   ...                       ...   
50000      58878.0                 BEXEY                  GO GETTA   
50001      43557.0             Roy Woods       Drama (feat. Drake)   
50002      39767.0                Berner  Lovin' Me (feat. Smiggz)   
50003      57944.0             The-Dream         Shawty Is Da Shit   
50004      63470.0     Naughty By Nature            Hip Hop Hooray   

       popularity  acousticness  danceability  duration_ms  energy  \
0            27.0       0.00468         0.652         -1.0   0.941   
1            31.0       0.01270         0.622     218293.0   0.890   
2            28.0       0.00306         0.620     215613.0   0.755   
3            34.0       0.02540         0.774     166875.0   0.700   
4            32.0       0.00465         0.638     222369.0   0.587   
...           ...           ...           ...          ...     ...   
50000        59.0       0.03340         0.913         -1.0   0.574   
50001        72.0       0.15700         0.709     251860.0   0.362   
50002        51.0       0.00597         0.693     189483.0   0.763   
50003        65.0       0.08310         0.782     262773.0   0.472   
50004        67.0       0.10200         0.862     267267.0   0.642   

       instrumentalness key  liveness  loudness   mode  speechiness  \
0               0.79200  A#     0.115    -5.201  Minor       0.0748   
1               0.95000   D     0.124    -7.043  Minor       0.0300   
2               0.01180  G#     0.534    -4.617  Major       0.0345   
3               0.00253  C#     0.157    -4.498  Major       0.2390   
4               0.90900  F#     0.157    -6.266  Major       0.0413   
...                 ...  ..       ...       ...    ...          ...   
50000           0.00000  C#     0.119    -7.022  Major       0.2980   
50001           0.00000   B     0.109    -9.814  Major       0.0550   
50002           0.00000   D     0.143    -5.443  Major       0.1460   
50003           0.00000   G     0.106    -5.016  Minor       0.0441   
50004           0.00000  F#     0.272   -13.652  Minor       0.1010   

                    tempo obtained_date  valence music_genre  
0                 100.889         4-Apr    0.759  Electronic  
1      115.00200000000001         4-Apr    0.531  Electronic  
2                 127.994         4-Apr    0.333  Electronic  
3                 128.014         4-Apr    0.270  Electronic  
4                 145.036         4-Apr    0.323  Electronic  
...                   ...           ...      ...         ...  
50000   98.02799999999999         4-Apr    0.330     Hip-Hop  
50001  122.04299999999999         4-Apr    0.113     Hip-Hop  
50002             131.079         4-Apr    0.395     Hip-Hop  
50003   75.88600000000001         4-Apr    0.354     Hip-Hop  
50004   99.20100000000001         4-Apr    0.765     Hip-Hop  

[50005 rows x 18 columns]

In [3]:
music.info()
music.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50005 entries, 0 to 50004
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   instance_id       50000 non-null  float64
 1   artist_name       50000 non-null  object 
 2   track_name        50000 non-null  object 
 3   popularity        50000 non-null  float64
 4   acousticness      50000 non-null  float64
 5   danceability      50000 non-null  float64
 6   duration_ms       50000 non-null  float64
 7   energy            50000 non-null  float64
 8   instrumentalness  50000 non-null  float64
 9   key               50000 non-null  object 
 10  liveness          50000 non-null  float64
 11  loudness          50000 non-null  float64
 12  mode              50000 non-null  object 
 13  speechiness       50000 non-null  float64
 14  tempo             50000 non-null  object 
 15  obtained_date     50000 non-null  object 
 16  valence           50000 non-null  float6

instance_id    popularity  acousticness  danceability   duration_ms  \
count  50000.000000  50000.000000  50000.000000  50000.000000  5.000000e+04   
mean   55888.396360     44.220420      0.306383      0.558241  2.212526e+05   
std    20725.256253     15.542008      0.341340      0.178632  1.286720e+05   
min    20002.000000      0.000000      0.000000      0.059600 -1.000000e+00   
25%    37973.500000     34.000000      0.020000      0.442000  1.748000e+05   
50%    55913.500000     45.000000      0.144000      0.568000  2.192810e+05   
75%    73863.250000     56.000000      0.552000      0.687000  2.686122e+05   
max    91759.000000     99.000000      0.996000      0.986000  4.830606e+06   

             energy  instrumentalness      liveness      loudness  \
count  50000.000000      50000.000000  50000.000000  50000.000000   
mean       0.599755          0.181601      0.193896     -9.133761   
std        0.264559          0.325409      0.161637      6.162990   
min        0.000792          0.000000      0.009670    -47.046000   
25%        0.433000          0.000000      0.096900    -10.860000   
50%        0.643000          0.000158      0.126000     -7.276500   
75%        0.815000          0.155000      0.244000     -5.173000   
max        0.999000          0.996000      1.000000      3.744000   

        speechiness       valence  
count  50000.000000  50000.000000  
mean       0.093586      0.456264  
std        0.101373      0.247119  
min        0.022300      0.000000  
25%        0.036100      0.257000  
50%        0.048900      0.448000  
75%        0.098525      0.648000  
max        0.942000      0.992000

In the dataframe, there exists 50005 rows. However, based on the description, only 50000 Non-null counts. Which means there are 5 rows with Null values. We will do the amendment in Data Cleaning.

# 2) Data Cleaning

### 2a) Check for any Null Values and remove.

Since we know there are 5 rows. Following code, we will identify the which rows consists Null value and remove specifically.

In [4]:
music[music.isnull().any(axis=1)]

instance_id artist_name track_name  popularity  acousticness  \
10000          NaN         NaN        NaN         NaN           NaN   
10001          NaN         NaN        NaN         NaN           NaN   
10002          NaN         NaN        NaN         NaN           NaN   
10003          NaN         NaN        NaN         NaN           NaN   
10004          NaN         NaN        NaN         NaN           NaN   

       danceability  duration_ms  energy  instrumentalness  key  liveness  \
10000           NaN          NaN     NaN               NaN  NaN       NaN   
10001           NaN          NaN     NaN               NaN  NaN       NaN   
10002           NaN          NaN     NaN               NaN  NaN       NaN   
10003           NaN          NaN     NaN               NaN  NaN       NaN   
10004           NaN          NaN     NaN               NaN  NaN       NaN   

       loudness mode  speechiness tempo obtained_date  valence music_genre  
10000       NaN  NaN          NaN   NaN           NaN      NaN         NaN  
10001       NaN  NaN          NaN   NaN           NaN      NaN         NaN  
10002       NaN  NaN          NaN   NaN           NaN      NaN         NaN  
10003       NaN  NaN          NaN   NaN           NaN      NaN         NaN  
10004       NaN  NaN          NaN   NaN           NaN      NaN         NaN

After identify its row 10000-10004, lets remove them accordingly.

In [5]:
music.drop([10000, 10001, 10002, 10003, 10004], inplace = True)

Next, we would like to check for any duplication rows in our dataset.

In [6]:
music[music.duplicated()]

Empty DataFrame
Columns: [instance_id, artist_name, track_name, popularity, acousticness, danceability, duration_ms, energy, instrumentalness, key, liveness, loudness, mode, speechiness, tempo, obtained_date, valence, music_genre]
Index: []

There isn't, we will move on.

In the dataset, we have also spotted some rows under column "Duration_ms" containing values of -1. This is invalid value, a song duration would never be -1, which may produce errors in our findings, hence we remove them

In [7]:
music.drop(music[music['duration_ms'] == -1.0].index, inplace = True)
music

instance_id           artist_name                track_name  \
1          46652.0  Thievery Corporation          The Shining Path   
2          30097.0        Dillon Francis                 Hurricane   
3          62177.0              Dubloadz                     Nitro   
4          24907.0           What So Not          Divide & Conquer   
5          89064.0            Axel Boman                     Hello   
...            ...                   ...                       ...   
49999      28408.0          Night Lovell               Barbie Doll   
50001      43557.0             Roy Woods       Drama (feat. Drake)   
50002      39767.0                Berner  Lovin' Me (feat. Smiggz)   
50003      57944.0             The-Dream         Shawty Is Da Shit   
50004      63470.0     Naughty By Nature            Hip Hop Hooray   

       popularity  acousticness  danceability  duration_ms  energy  \
1            31.0       0.01270         0.622     218293.0   0.890   
2            28.0       0.00306         0.620     215613.0   0.755   
3            34.0       0.02540         0.774     166875.0   0.700   
4            32.0       0.00465         0.638     222369.0   0.587   
5            47.0       0.00523         0.755     519468.0   0.731   
...           ...           ...           ...          ...     ...   
49999        56.0       0.13300         0.849     237667.0   0.660   
50001        72.0       0.15700         0.709     251860.0   0.362   
50002        51.0       0.00597         0.693     189483.0   0.763   
50003        65.0       0.08310         0.782     262773.0   0.472   
50004        67.0       0.10200         0.862     267267.0   0.642   

       instrumentalness key  liveness  loudness   mode  speechiness  \
1              0.950000   D     0.124    -7.043  Minor       0.0300   
2              0.011800  G#     0.534    -4.617  Major       0.0345   
3              0.002530  C#     0.157    -4.498  Major       0.2390   
4              0.909000  F#     0.157    -6.266  Major       0.0413   
5              0.854000   D     0.216   -10.517  Minor       0.0412   
...                 ...  ..       ...       ...    ...          ...   
49999          0.000008   C     0.296    -7.195  Major       0.0516   
50001          0.000000   B     0.109    -9.814  Major       0.0550   
50002          0.000000   D     0.143    -5.443  Major       0.1460   
50003          0.000000   G     0.106    -5.016  Minor       0.0441   
50004          0.000000  F#     0.272   -13.652  Minor       0.1010   

                    tempo obtained_date  valence music_genre  
1      115.00200000000001         4-Apr    0.531  Electronic  
2                 127.994         4-Apr    0.333  Electronic  
3                 128.014         4-Apr    0.270  Electronic  
4                 145.036         4-Apr    0.323  Electronic  
5                       ?         4-Apr    0.614  Electronic  
...                   ...           ...      ...         ...  
49999              99.988         4-Apr    0.629     Hip-Hop  
50001  122.04299999999999         4-Apr    0.113     Hip-Hop  
50002             131.079         4-Apr    0.395     Hip-Hop  
50003   75.88600000000001         4-Apr    0.354     Hip-Hop  
50004   99.20100000000001         4-Apr    0.765     Hip-Hop  

[45061 rows x 18 columns]

Besides duration_ms, "tempo" turns out to have invalid value as well "?", These values need to either be removed or replaced with median or mean value. If not errors may occur during our findings. 

In [8]:
music[music["tempo"] == "?"]

instance_id artist_name               track_name  popularity  \
5          89064.0  Axel Boman                    Hello        47.0   
32         25836.0    PEEKABOO                  Arrival        45.0   
36         55617.0  Wax Tailor       The Games You Play        45.0   
39         69685.0        Dahu                   Vessel        37.0   
41         83854.0       UNKLE              Lonely Soul        51.0   
...            ...         ...                      ...         ...   
49918      63058.0    Big Sean           Bigger Than Me        58.0   
49964      53387.0  Millonario          Rayas de Patrón        59.0   
49967      76585.0   MadeinTYO  I Want (feat. 2 Chainz)        62.0   
49976      79654.0    Big Sean   Sunday Morning Jetpack        52.0   
49977      63945.0   Nate Dogg               Music & Me        58.0   

       acousticness  danceability  duration_ms  energy  instrumentalness key  \
5           0.00523         0.755     519468.0   0.731          0.854000   D   
32          0.02330         0.729     274286.0   0.869          0.585000   F   
36          0.04780         0.646     253333.0   0.649          0.002520   G   
39          0.20300         0.769     429941.0   0.551          0.882000  A#   
41          0.07760         0.475     533013.0   0.591          0.054900   A   
...             ...           ...          ...     ...               ...  ..   
49918       0.29600         0.379     292520.0   0.644          0.000000  A#   
49964       0.08470         0.929     215200.0   0.737          0.000000  G#   
49967       0.17900         0.860     233293.0   0.625          0.000136   D   
49976       0.70000         0.462     225067.0   0.741          0.000000  A#   
49977       0.10500         0.905     240627.0   0.414          0.000366  G#   

       liveness  loudness   mode  speechiness tempo obtained_date  valence  \
5        0.2160   -10.517  Minor       0.0412     ?         4-Apr   0.6140   
32       0.0944    -7.569  Minor       0.1290     ?         4-Apr   0.1380   
36       0.3530    -5.259  Major       0.1170     ?         4-Apr   0.7270   
39       0.1090   -12.520  Minor       0.0408     ?         4-Apr   0.0618   
41       0.1090    -7.854  Minor       0.0374     ?         4-Apr   0.1230   
...         ...       ...    ...          ...   ...           ...      ...   
49918    0.3130    -7.933  Minor       0.1120     ?         5-Apr   0.1440   
49964    0.8610    -6.424  Major       0.1020     ?         4-Apr   0.8860   
49967    0.3000    -6.922  Major       0.1050     ?         4-Apr   0.2210   
49976    0.3400    -8.008  Minor       0.1620     ?         4-Apr   0.5890   
49977    0.0914    -8.112  Minor       0.0615     ?         4-Apr   0.7580   

      music_genre  
5      Electronic  
32     Electronic  
36     Electronic  
39     Electronic  
41     Electronic  
...           ...  
49918     Hip-Hop  
49964     Hip-Hop  
49967     Hip-Hop  
49976     Hip-Hop  
49977     Hip-Hop  

[4501 rows x 18 columns]

There exists 4501 rows with tempo value "?", this problem is significant. Hence, need to be removed.

In [9]:
music.drop(music[music['tempo'] == "?"].index, inplace = True)
music

instance_id           artist_name                track_name  \
1          46652.0  Thievery Corporation          The Shining Path   
2          30097.0        Dillon Francis                 Hurricane   
3          62177.0              Dubloadz                     Nitro   
4          24907.0           What So Not          Divide & Conquer   
6          43760.0        Jordan Comolli                     Clash   
...            ...                   ...                       ...   
49999      28408.0          Night Lovell               Barbie Doll   
50001      43557.0             Roy Woods       Drama (feat. Drake)   
50002      39767.0                Berner  Lovin' Me (feat. Smiggz)   
50003      57944.0             The-Dream         Shawty Is Da Shit   
50004      63470.0     Naughty By Nature            Hip Hop Hooray   

       popularity  acousticness  danceability  duration_ms  energy  \
1            31.0       0.01270         0.622     218293.0   0.890   
2            28.0       0.00306         0.620     215613.0   0.755   
3            34.0       0.02540         0.774     166875.0   0.700   
4            32.0       0.00465         0.638     222369.0   0.587   
6            46.0       0.02890         0.572     214408.0   0.803   
...           ...           ...           ...          ...     ...   
49999        56.0       0.13300         0.849     237667.0   0.660   
50001        72.0       0.15700         0.709     251860.0   0.362   
50002        51.0       0.00597         0.693     189483.0   0.763   
50003        65.0       0.08310         0.782     262773.0   0.472   
50004        67.0       0.10200         0.862     267267.0   0.642   

       instrumentalness key  liveness  loudness   mode  speechiness  \
1              0.950000   D     0.124    -7.043  Minor       0.0300   
2              0.011800  G#     0.534    -4.617  Major       0.0345   
3              0.002530  C#     0.157    -4.498  Major       0.2390   
4              0.909000  F#     0.157    -6.266  Major       0.0413   
6              0.000008   B     0.106    -4.294  Major       0.3510   
...                 ...  ..       ...       ...    ...          ...   
49999          0.000008   C     0.296    -7.195  Major       0.0516   
50001          0.000000   B     0.109    -9.814  Major       0.0550   
50002          0.000000   D     0.143    -5.443  Major       0.1460   
50003          0.000000   G     0.106    -5.016  Minor       0.0441   
50004          0.000000  F#     0.272   -13.652  Minor       0.1010   

                    tempo obtained_date  valence music_genre  
1      115.00200000000001         4-Apr    0.531  Electronic  
2                 127.994         4-Apr    0.333  Electronic  
3                 128.014         4-Apr    0.270  Electronic  
4                 145.036         4-Apr    0.323  Electronic  
6                 149.995         4-Apr    0.230  Electronic  
...                   ...           ...      ...         ...  
49999              99.988         4-Apr    0.629     Hip-Hop  
50001  122.04299999999999         4-Apr    0.113     Hip-Hop  
50002             131.079         4-Apr    0.395     Hip-Hop  
50003   75.88600000000001         4-Apr    0.354     Hip-Hop  
50004   99.20100000000001         4-Apr    0.765     Hip-Hop  

[40560 rows x 18 columns]

The reason why we are removing invalid values instead of replacing it with median or mean values is because simplifies the data and makes it easier to work with and bias may happen if the mean or median value is inappropriate or if the distribution of the data is skewed.

In [10]:
#Doing a final check on the dataset
pd.isnull(music).sum() #Check for any remaining Null Values 

instance_id         0
artist_name         0
track_name          0
popularity          0
acousticness        0
danceability        0
duration_ms         0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
obtained_date       0
valence             0
music_genre         0
dtype: int64

In [11]:
music[music["tempo"] == "?"]

Empty DataFrame
Columns: [instance_id, artist_name, track_name, popularity, acousticness, danceability, duration_ms, energy, instrumentalness, key, liveness, loudness, mode, speechiness, tempo, obtained_date, valence, music_genre]
Index: []

In [12]:
music[music["duration_ms"] == -1.0]

Empty DataFrame
Columns: [instance_id, artist_name, track_name, popularity, acousticness, danceability, duration_ms, energy, instrumentalness, key, liveness, loudness, mode, speechiness, tempo, obtained_date, valence, music_genre]
Index: []

# Remove unwanted columns

In the dataset, there exist some unwanted column such as "obtained_date", "instance_id", "track_name" and "artist_name". These columns do not have any relation with the song genre, a date, id and track_name doesn't determine a song's genre and wouldn't be able to help us with the prediction for song genre. And same artist may produce a different types of genre, we cannot interpret anything based on the artist name. Hence we are removing these columns."

In [13]:
music = music.drop(["instance_id", "track_name", "obtained_date", "artist_name"], axis = 1)
music.head()

popularity  acousticness  danceability  duration_ms  energy  \
1        31.0       0.01270         0.622     218293.0   0.890   
2        28.0       0.00306         0.620     215613.0   0.755   
3        34.0       0.02540         0.774     166875.0   0.700   
4        32.0       0.00465         0.638     222369.0   0.587   
6        46.0       0.02890         0.572     214408.0   0.803   

   instrumentalness key  liveness  loudness   mode  speechiness  \
1          0.950000   D     0.124    -7.043  Minor       0.0300   
2          0.011800  G#     0.534    -4.617  Major       0.0345   
3          0.002530  C#     0.157    -4.498  Major       0.2390   
4          0.909000  F#     0.157    -6.266  Major       0.0413   
6          0.000008   B     0.106    -4.294  Major       0.3510   

                tempo  valence music_genre  
1  115.00200000000001    0.531  Electronic  
2             127.994    0.333  Electronic  
3             128.014    0.270  Electronic  
4             145.036    0.323  Electronic  
6             149.995    0.230  Electronic

The dataset index was reset since 5 rows were removed earlier 

In [16]:
music.reset_index(inplace = True)

After non-predictive features are removed, here is the updated dataset after this initial clean-up.

In [17]:
music

level_0  index  popularity  acousticness  danceability  duration_ms  \
0            0      1        31.0       0.01270         0.622     218293.0   
1            1      2        28.0       0.00306         0.620     215613.0   
2            2      3        34.0       0.02540         0.774     166875.0   
3            3      4        32.0       0.00465         0.638     222369.0   
4            4      6        46.0       0.02890         0.572     214408.0   
...        ...    ...         ...           ...           ...          ...   
40555    40555  49999        56.0       0.13300         0.849     237667.0   
40556    40556  50001        72.0       0.15700         0.709     251860.0   
40557    40557  50002        51.0       0.00597         0.693     189483.0   
40558    40558  50003        65.0       0.08310         0.782     262773.0   
40559    40559  50004        67.0       0.10200         0.862     267267.0   

       energy  instrumentalness key  liveness  loudness   mode  speechiness  \
0       0.890          0.950000   D     0.124    -7.043  Minor       0.0300   
1       0.755          0.011800  G#     0.534    -4.617  Major       0.0345   
2       0.700          0.002530  C#     0.157    -4.498  Major       0.2390   
3       0.587          0.909000  F#     0.157    -6.266  Major       0.0413   
4       0.803          0.000008   B     0.106    -4.294  Major       0.3510   
...       ...               ...  ..       ...       ...    ...          ...   
40555   0.660          0.000008   C     0.296    -7.195  Major       0.0516   
40556   0.362          0.000000   B     0.109    -9.814  Major       0.0550   
40557   0.763          0.000000   D     0.143    -5.443  Major       0.1460   
40558   0.472          0.000000   G     0.106    -5.016  Minor       0.0441   
40559   0.642          0.000000  F#     0.272   -13.652  Minor       0.1010   

                    tempo  valence music_genre  
0      115.00200000000001    0.531  Electronic  
1                 127.994    0.333  Electronic  
2                 128.014    0.270  Electronic  
3                 145.036    0.323  Electronic  
4                 149.995    0.230  Electronic  
...                   ...      ...         ...  
40555              99.988    0.629     Hip-Hop  
40556  122.04299999999999    0.113     Hip-Hop  
40557             131.079    0.395     Hip-Hop  
40558   75.88600000000001    0.354     Hip-Hop  
40559   99.20100000000001    0.765     Hip-Hop  

[40560 rows x 16 columns]

In [15]:
music.to_csv('cleaned_music.csv')

Converting the cleaned music dataset to new csv file for the next step 3) Data Exploratory Analysis.